File: fixed/utils/format.js
/**
 * Format a number for the calculator display.
 * Adds commas for thousands, handles special values.
 * @param {string} value - The display string
 * @returns {string}
 */
export function formatDisplay(value) {
  if (value === '' || value === null || value === undefined) return '0'
  if (value === 'Error') return 'Error'

  const str = String(value)

  // Don't format if user is still typing a decimal
  if (str.endsWith('.')) return str

  const num = parseFloat(str)
  if (isNaN(num)) return 'Error'
  if (!isFinite(num)) return 'Error'

  // If it has a decimal part, limit to 10 decimal places
  if (str.includes('.') || (num % 1 !== 0)) {
    const formatted = parseFloat(num.toPrecision(12))
    return String(formatted)
  }

  return String(num)
}

/**
 * Format a history entry string.
 * @param {number|string} a - Left operand
 * @param {string} op - Operator symbol
 * @param {number|string} b - Right operand
 * @param {number|string} result - Calculation result
 * @returns {string}
 */
export function formatHistoryEntry(a, op, b, result) {
  const opSymbol = { '+': '+', '-': '\u2212', '*': '\u00d7', '/': '\u00f7' }[op] || op
  return `${a} ${opSymbol} ${b} = ${result}`
}

File: buggy-java/src/main/java/bughuntbingo/utils/Format.java
package bughuntbingo.utils;

import java.util.Map;

public class Format {

    /**
     * Format a number for the calculator display.
     * Adds commas for thousands, handles special values.
     */
    public static String formatDisplay(String value) {
        if (value == null || value.isEmpty()) return "0";
        if (value.equals("Error")) return "Error";
        if (value.endsWith(".")) return value;

        try {
            double num = Double.parseDouble(value);
            if (Double.isNaN(num) || Double.isInfinite(num)) return "Error";

            if (value.contains(".") || num % 1 != 0) {
                double formatted = Double.parseDouble(String.format("%.12g", num));
                String s = Double.toString(formatted);
                if (s.endsWith(".0")) s = s.substring(0, s.length() - 2);
                return s;
            }
            return String.valueOf((long) num);
        } catch (NumberFormatException e) {
            return "Error";
        }
    }

    public static String formatHistoryEntry(double a, String op, double b, Object result) {
        Map<String, String> symbols = Map.of(
            "+", "+", "-", "\u2212", "*", "\u00d7", "/", "\u00f7"
        );
        String sym = symbols.getOrDefault(op, op);
        String aStr = a % 1 == 0 ? String.valueOf((long) a) : String.valueOf(a);
        String bStr = b % 1 == 0 ? String.valueOf((long) b) : String.valueOf(b);
        return aStr + " " + sym + " " + bStr + " = " + result;
    }
}
